### PR TITLE
Prevent retargeting of events from shadow descendants of light children.

### PR DESCRIPTION
**Pull request by bicknellr**
(fixes PolymerElements/paper-menu#89)

### CODE REVIEW

**Comment by ktiedt** [comment]
This does indeed appear to fix both issues here:

https://github.com/PolymerElements/iron-menu-behavior/issues/49
https://github.com/PolymerElements/paper-menu/issues/89

It actually makes keyboard a11y 100x better (its still quirky, but better) for paper-menu/sub-menu

Thanks @bicknellr!

**Comment by ktiedt** [comment]
Just following up, can this get merged in or is there more testing being done before this lands?

**Comment by bicknellr** [comment]
(ping @cdata)

File: test/test-elements.html [hunk 1]
```diff
@@ -89,3 +89,19 @@
   });
 
 </script>
+
+<dom-module id="control-with-shadow-input">
+  <template>
+    <input id="shadowInput">
```
Comment (notwaldorf) [inline test/test-elements.html]: isn't this the same as `nested-focusable`? Can you just add `IronButtonState` to that one?

Comment (bicknellr) [inline test/test-elements.html]: Yup, totally missed that. I think I grabbed `IronButtonState` when copying the `test-light-dom` element above but this is really just part of `IronControlState`, do you think I still need to add it?

**Comment by ktiedt** [comment]
Ping?

File: iron-control-state.html [hunk 2]
```diff
@@ -73,12 +73,19 @@
...
-      } else if (!this.shadowRoot && !this.isLightDescendant(event.target)) {
-        this.fire(event.type, {sourceEvent: event}, {
-          node: this,
-          bubbles: event.bubbles,
-          cancelable: event.cancelable
-        });
+      } else if (!this.shadowRoot && this.root) {
+        var ownerRoot = Polymer.dom(event.target).getOwnerRoot();
+
+        while (ownerRoot && ownerRoot !== this.root) {
```
Comment (notwaldorf) [inline iron-control-state.html]: If you're just checking if this is in your shadow dom, can you use `isLocalDescendant`? https://github.com/Polymer/polymer/blob/master/src/standard/utils.html#L428

File: iron-control-state.html [hunk 3]
```diff
@@ -73,12 +73,19 @@
 
       if (event.target === this) {
         this._setFocused(event.type === 'focus');
-      } else if (!this.shadowRoot && !this.isLightDescendant(event.target)) {
```
Comment (cdata) [inline iron-control-state.html]: I think that the real bug here may be in this line. I suspect that `this.isLightDescendant(event.target)` needs to be changed to `this.isLightDescendant(Polymer.dom(event).localTarget)`.

Comment (bicknellr) [inline iron-control-state.html]: Oh, or even `Polymer.dom(event).localTarget === this`, to match the spec text I thought didn't exist earlier.

File: iron-control-state.html (hunk 2 above)
Comment (bicknellr) [inline iron-control-state.html]: Yeah, in shadow but also if it was a descendant of anything in there too. This whole check should be unnecessary after making the change to the outer condition though.

File: iron-control-state.html (hunk 3 above)
Comment (bicknellr) [inline iron-control-state.html]: pushed

**Comment by tjsavage** [comment]
Is this ready to be reviewed/merged?

File: iron-control-state.html [hunk 4]
```diff
@@ -73,9 +73,8 @@
...
+      } else if (!this.shadowRoot && Polymer.dom(event).localTarget === this) {
         this.fire(event.type, {sourceEvent: event}, {
-          node: this,
```
Comment (notwaldorf) [inline iron-control-state.html]: Why isn't this needed anymore?

Comment (bicknellr) [inline iron-control-state.html]: `this.fire(..., { node: this, ...}` is redundant, right?

**Comment by ktiedt** [comment]
Friendly weekly ping, we have another user being bit by this bug now.

File: iron-control-state.html (hunk 3 above)
Comment (notwaldorf) [inline iron-control-state.html]: Not sure that's correct -- isn't that too strict? You want this code to be ran for your direct shadow children, don't you?

/cc @cdata

Comment (bicknellr) [inline iron-control-state.html]: Yes, since the event listener is on the host rather than the ShadowRoot, `.localTarget === this` after retargeting.

**Comment by ktiedt** [comment]
[Not so] friendly ping? :disappointed:

File: iron-control-state.html (hunk 3 above)
Comment (bicknellr) [inline iron-control-state.html]: (Changed to @cdata's initial suggestion.)

**Comment by cdata** [comment]
LGTM 👍

**Pull request merged**